File: flutter_inappwebview_android/android/src/main/kotlin/dev/nosferatu500/inappwebview/in_app_browser/InAppBrowserSettings.kt
package dev.nosferatu500.inappwebview.in_app_browser

import dev.nosferatu500.inappwebview.ISettings
import dev.nosferatu500.inappwebview.R

class InAppBrowserSettings : ISettings<InAppBrowserActivity> {

  @JvmField var hidden: Boolean = false
  @JvmField var hideToolbarTop: Boolean = false
  @JvmField var toolbarTopBackgroundColor: String? = null
  @JvmField var toolbarTopFixedTitle: String? = null
  @JvmField var hideUrlBar: Boolean = false
  @JvmField var hideProgressBar: Boolean = false

  @JvmField var hideTitleBar: Boolean = false
  @JvmField var closeOnCannotGoBack: Boolean = true
  @JvmField var allowGoBackWithBackButton: Boolean = true
  @JvmField var shouldCloseOnBackButtonPressed: Boolean = false
  @JvmField var hideDefaultMenuItems: Boolean = false

  override fun parse(settings: Map<String, Any?>): InAppBrowserSettings {
    for ((key, value) in settings) {
      if (value == null) {
        continue
      }
      when (key) {
        "hidden" -> hidden = value as Boolean
        "hideToolbarTop" -> hideToolbarTop = value as Boolean
        "toolbarTopBackgroundColor" -> toolbarTopBackgroundColor = value as String
        "toolbarTopFixedTitle" -> toolbarTopFixedTitle = value as String
        "hideUrlBar" -> hideUrlBar = value as Boolean
        "hideTitleBar" -> hideTitleBar = value as Boolean
        "closeOnCannotGoBack" -> closeOnCannotGoBack = value as Boolean
        "hideProgressBar" -> hideProgressBar = value as Boolean
        "allowGoBackWithBackButton" -> allowGoBackWithBackButton = value as Boolean
        "shouldCloseOnBackButtonPressed" -> shouldCloseOnBackButtonPressed = value as Boolean
        "hideDefaultMenuItems" -> hideDefaultMenuItems = value as Boolean
      }
    }
    return this
  }

  override fun toMap(): MutableMap<String, Any?> = hashMapOf(
    "hidden" to hidden,
    "hideToolbarTop" to hideToolbarTop,
    "toolbarTopBackgroundColor" to toolbarTopBackgroundColor,
    "toolbarTopFixedTitle" to toolbarTopFixedTitle,
    "hideUrlBar" to hideUrlBar,
    "hideTitleBar" to hideTitleBar,
    "closeOnCannotGoBack" to closeOnCannotGoBack,
    "hideProgressBar" to hideProgressBar,
    "allowGoBackWithBackButton" to allowGoBackWithBackButton,
    "shouldCloseOnBackButtonPressed" to shouldCloseOnBackButtonPressed,
    "hideDefaultMenuItems" to hideDefaultMenuItems
  )

  override fun getRealSettings(obj: InAppBrowserActivity): MutableMap<String, Any?> {
    val realSettings = toMap()
    realSettings["hidden"] = obj.isHidden
    realSettings["hideToolbarTop"] = obj.actionBar?.isShowing != true
    realSettings["hideUrlBar"] =
      obj.menu?.findItem(R.id.menu_search)?.isVisible != true
    realSettings["hideProgressBar"] = obj.progressBar.let { it == null || it.max == 0 }
    return realSettings
  }

  companion object {
    const val LOG_TAG = "InAppBrowserSettings"
  }
}

File: flutter_inappwebview_platform_interface/lib/src/in_app_browser/in_app_browser_settings.dart
import 'dart:ui';

import 'package:flutter/foundation.dart';
import 'package:flutter_inappwebview_internal_annotations/flutter_inappwebview_internal_annotations.dart';

import '../types/modal_presentation_style.dart';
import '../types/modal_transition_style.dart';
import '../util.dart';

import '../in_app_webview/in_app_webview_settings.dart';

import '../types/enum_method.dart';

part 'in_app_browser_settings.g.dart';

///Class that represents the settings that can be used for an [InAppBrowser] instance.
class InAppBrowserClassSettings {
  ///Browser settings.
  late InAppBrowserSettings browserSettings;

  ///WebView settings.
  late InAppWebViewSettings webViewSettings;

  InAppBrowserClassSettings({
    InAppBrowserSettings? browserSettings,
    InAppWebViewSettings? webViewSettings,
  }) {
    this.browserSettings = browserSettings ?? InAppBrowserSettings();
    this.webViewSettings = webViewSettings ?? InAppWebViewSettings();
  }

  Map<String, dynamic> toMap() {
    Map<String, dynamic> options = {};

    options.addAll(browserSettings.toMap());
    options.addAll(webViewSettings.toMap());

    return options;
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }

  @override
  String toString() {
    return toMap().toString();
  }

  factory InAppBrowserClassSettings.fromMap(
    Map<String, dynamic> options, {
    InAppBrowserClassSettings? instance,
    EnumMethod? enumMethod,
  }) {
    instance ??= InAppBrowserClassSettings();
    instance.browserSettings =
        InAppBrowserSettings.fromMap(options, enumMethod: enumMethod) ??
        InAppBrowserSettings();
    instance.webViewSettings =
        InAppWebViewSettings.fromMap(options, enumMethod: enumMethod) ??
        InAppWebViewSettings();
    return instance;
  }

  InAppBrowserClassSettings copy() {
    return InAppBrowserClassSettings.fromMap(toMap());
  }
}

///{@template flutter_inappwebview_platform_interface.InAppWebViewSettings}
///This class represents all [InAppBrowser] settings available.
///{@endtemplate}
@ExchangeableObject(copyMethod: true)
@SupportedPlatforms(platforms: [AndroidPlatform(), IOSPlatform()])
class InAppBrowserSettings_ {
  ///Set to `true` to create the browser and load the page, but not show it. Omit or set to `false` to have the browser open and load normally.
  ///The default value is `false`.
  @SupportedPlatforms(platforms: [AndroidPlatform(), IOSPlatform()])
  bool? hidden;

  ///Set to `true` to hide the toolbar at the top of the WebView. The default value is `false`.
  @SupportedPlatforms(platforms: [AndroidPlatform(), IOSPlatform()])
  bool? hideToolbarTop;

  ///Set the custom background color of the toolbar at the top.
  @SupportedPlatforms(platforms: [AndroidPlatform(), IOSPlatform()])
  Color_? toolbarTopBackgroundColor;

  ///Set to `true` to hide the url bar on the toolbar at the top. The default value is `false`.
  @SupportedPlatforms(platforms: [AndroidPlatform(), IOSPlatform()])
  bool? hideUrlBar;

  ///Set to `true` to hide the progress bar when the WebView is loading a page. The default value is `false`.
  @SupportedPlatforms(platforms: [AndroidPlatform(), IOSPlatform()])
  bool? hideProgressBar;

  ///Set to `true` to hide the default menu items. The default value is `false`.
  @SupportedPlatforms(platforms: [AndroidPlatform(), IOSPlatform()])
  bool? hideDefaultMenuItems;

  ///Set to `true` if you want the title should be displayed. The default value is `false`.
  @SupportedPlatforms(platforms: [AndroidPlatform()])
  bool? hideTitleBar;

  ///Set the action bar's title.
  @SupportedPlatforms(platforms: [AndroidPlatform()])
  String? toolbarTopFixedTitle;

  ///Set to `false` to not close the InAppBrowser when the user click on the Android back button and the WebView cannot go back to the history. The default value is `true`.
  @SupportedPlatforms(platforms: [AndroidPlatform()])
  bool? closeOnCannotGoBack;

  ///Set to `false` to block the InAppBrowser WebView going back when the user click on the Android back button. The default value is `true`.
  @SupportedPlatforms(platforms: [AndroidPlatform()])
  bool? allowGoBackWithBackButton;

  ///Set to `true` to close the InAppBrowser when the user click on the Android back button. The default value is `false`.
  @SupportedPlatforms(platforms: [AndroidPlatform()])
  bool? shouldCloseOnBackButtonPressed;

  ///Set to `true` to set the toolbar at the top translucent. The default value is `true`.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  bool? toolbarTopTranslucent;

  ///Set the tint color to apply to the navigation bar background.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  Color_? toolbarTopBarTintColor;

  ///Set the tint color to apply to the navigation items and bar button items.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  Color_? toolbarTopTintColor;

  ///Set to `true` to hide the toolbar at the bottom of the WebView. The default value is `false`.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  bool? hideToolbarBottom;

  ///Set the custom background color of the toolbar at the bottom.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  Color_? toolbarBottomBackgroundColor;

  ///Set the tint color to apply to the bar button items.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  Color_? toolbarBottomTintColor;

  ///Set to `true` to set the toolbar at the bottom translucent. The default value is `true`.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  bool? toolbarBottomTranslucent;

  ///Set the custom text for the close button.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  String? closeButtonCaption;

  ///Set the custom color for the close button.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  Color_? closeButtonColor;

  ///Set to `true` to hide the close button. The default value is `false`.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  bool? hideCloseButton;

  ///Set the custom color for the menu button.
  @SupportedPlatforms(platforms: [IOSPlatform()])
  Color_? menuButtonColor;

  ///Set the custom modal presentation style when presenting the WebView. The default value is [ModalPresentationStyle.FULL_SCREEN].
  @SupportedPlatforms(platforms: [IOSPlatform()])
  ModalPresentationStyle_? presentationStyle;

  ///Set to the custom transition style when presenting the WebView. The default value is [ModalTransitionStyle.COVER_VERTICAL].
  @SupportedPlatforms(platforms: [IOSPlatform()])
  ModalTransitionStyle_? transitionStyle;

  InAppBrowserSettings_({
    this.hidden = false,
    this.hideToolbarTop = false,
    this.toolbarTopBackgroundColor,
    this.hideUrlBar = false,
    this.hideProgressBar = false,
    this.hideDefaultMenuItems = false,
    this.toolbarTopTranslucent = true,
    this.toolbarTopTintColor,
    this.hideToolbarBottom = false,
    this.toolbarBottomBackgroundColor,
    this.toolbarBottomTintColor,
    this.toolbarBottomTranslucent = true,
    this.closeButtonCaption,
    this.closeButtonColor,
    this.hideCloseButton = false,
    this.menuButtonColor,
    this.presentationStyle = ModalPresentationStyle_.FULL_SCREEN,
    this.transitionStyle = ModalTransitionStyle_.COVER_VERTICAL,
    this.hideTitleBar = false,
    this.toolbarTopFixedTitle,
    this.closeOnCannotGoBack = true,
    this.allowGoBackWithBackButton = true,
    this.shouldCloseOnBackButtonPressed = false,
  });

  ///Check if the given [property] is supported by the [defaultTargetPlatform] or a specific [platform].
  static bool isPropertySupported(
    InAppBrowserSettingsProperty property, {
    TargetPlatform? platform,
  }) => _InAppBrowserSettingsPropertySupported.isPropertySupported(
    property,
    platform: platform,
  );
}
